Snowplow/Internal/Tracker/SPTracker.m

```diff
@@ -528,10 +528,19 @@ - (void)track:(SPEvent *)event {
 
 - (void)processEvent:(SPEvent *)event {
     SPTrackerEvent *trackerEvent = [[SPTrackerEvent alloc] initWithEvent:event];
+    [self transformEvent:trackerEvent];
     SPPayload *payload = [self payloadWithEvent:trackerEvent];
     [_emitter addPayloadToBuffer:payload];
 }
 
+- (void)transformEvent:(SPTrackerEvent *)event {
```

Comment:
This feels a little strange but I see why we have to do it given `SPEvent` vs `SPTrackerEvent`. 
I don't think we should change anything for this release but for future releases I can think of a couple of alternatives. They might not be any better than this route though, I'll let you decide that.

- It might be worth considering if we could find a way to set payload properties on an `SPEvent` that are "preferred" over any automatic properties that are created when initialising `SPTrackerEvent`. That way you could set `dtm` when creating the install event rather than having to do this swapping on the timestamps.
- Alternatively, maybe we could make a `SPApplicationInstall` event type and leverage the `beginProcessing` function?

Reply:
I think the problem with application_install comes from a misuse of "dtm" in this case. "dtm" represents the timestamp of the event creation and it shouldn't change. It's like rewrite the history, and it should part of a lower abstraction at a _tracker level_. I think the application_install should have its own timestamp in the payload  (lets say at the _application level_) which specifies the _first_run_ timestamp (as we can't really track the installation).

Unfortunately, using the true_timestamp was a hacky solution in order to keep the "dtm" pure and fully automatic.
I think your idea of "preferred" properties is a very good idea, from a pragmatic point of view, as it easily solve the problem but it doesn't represents the two levels mentioned above: tracker and application. I'm just scared of conceding flexibility that can drive to misuse of the fields like for the applicaton_install and "dtm" in the past.

The idea of using "beginProcessing" at the moment can't work because it's executed on the original event, which doesn't have access to the "dtm" as it's a tracker defined field. It should be executed on the SPTrackerEvent but it would raise the problem of giving to the developer the full access to internal properties. Not a problem itself. It's just that we don't know who and how these fields are used and it can create doubts when we troubleshoot a customer issue.

I think "transformEvent" solution is not particularly elegant or smart but it can have a place in the larger picture. It can be the phase of the event processing where we _transform/block_ the event based on some rules. The so long debated edge analytics. Also if it would raise concerns about the performances.

Anyway, I'm open to all the possible solutions, just we need to make them coherent with the rest of the internal. Definitely something to consider when we split the old SPTracker in pieces.

Another bit of brain dump on true_timestamp: I think that passing that directly to the derived_timestamp is dangerous. Probably it wasn't designed for the mobile devices. IMO a better solution might be: `derived = collector - (sent - true)` like for the dvce_tmstamp.

Reply:
Good points. I actually think you've found the right solution there. Updating the application install schema to have its own time stamp!